### DISCUSSION TITLE
Set formatted number value to an Controlled input

### DISCUSSION THREAD

**Discussion by patrickjmeurer** [react-hook-form/react-hook-form#9161] [Q&A]
Hello Guys, i'm trying to set a formatted value (currency) to a MUI Textfield, but i got stucked.

Here is the component's code:
```javascript
<Controller
    control={control}
    name="moneyField"
    render={({ field: { onChange, value } }) => (
      <TextField
        // type="number"
        label={label}
        value={value}
        placeholder={placeholder}
        fullWidth
        onChange={({ target: { value: targetValue } }) => {
          onChange(formatMoney(targetValue));
        }}
      />
    )}
  />

const formatMoney = (price: number | string) => {
  const newValue = price.toString().replace('R$', '').replace(',', '.');

  const result = new Intl.NumberFormat('pt-BR', {
    minimumFractionDigits: 2,
    style: 'currency',
    currency: 'BRL',
  }).format(parseFloat(newValue));

  return result;
};
```

When i type some numbers, i got stucked on the first number i typed:
![chrome_HedCsxBe80](https://user-images.githubusercontent.com/8931135/194443938-5f6d4376-e583-43e1-9482-2f50cbe759ef.png)

Someone can help me, please D=

Thanks!

**Comment by leapful** [comment]
@potreco Could you please provide a codesandbox url that can re-produce the issue ?

**Reply by patrickjmeurer** [reply]
@leapful, for sure!

https://codesandbox.io/s/quirky-dewdney-trpfi6

Thanks!

**Reply by leapful** [reply]
@potreco : I think you are confused on the behavior of input custom format. Event handler "onBlur" should be used to format the value number instead of "onChange" as it will be applied at the end of user typing journey, not during typing journey.

It's the common behavior when dealing with input custom format. Some examples can be found in https://dm4t2.github.io/vue-currency-input/playground.html

Working demo: https://codesandbox.io/s/competent-hellman-s2ofge?file=/src/MoneyInput.js

**Reply by patrickjmeurer** [reply]
Hello @leapful ,

I cannot use the onChange to mask the input? We need mask the value when the user put some digits, not when focus out. :(

Maybe use the `onKeyUp` functions will work?

Thanks!

**Reply by leapful** [reply]
@potreco

Not, it does not work like that. You are formatting the value on each key press, and the new format is not same length with input value therefore it always alter current selection cursor of input and causes many unexpected issues on typing.

For example: 

Type "1" => Formatted to "R$ 1.00" 
=> Now selection cursor is at last character "0" 
=> Type "2" => Value is now "R$ 1.002" 
=> Your function "formatMoney" will format "1.002" to "R$ 1.00" as the option fraction digit is 2

It's impossible to achieve your expectation without having a gap between user typing and formatting. You must use "onBlur" or at least debounce "formatMoney" until user ends typing.
